File: 540.有序数组中的单一元素.cpp
/*
 * @lc app=leetcode.cn id=540 lang=cpp
 *
 * 二分查找，双指针
 * 时间复杂度O(logN)
 * 
 * [540] 有序数组中的单一元素
 */

// @lc code=start
class Solution {
public:
    int singleNonDuplicate(vector<int> &nums)
    {
        int l = 0, r = nums.size()-1;
        while (l < r)
        {
            int mid = (l + r) / 4 * 2;
            if (nums[mid] == nums[mid + 1])
            {
                l = mid + 2;
            }
            else
            {
                r = mid;
            }
        }
        return nums[l];
    }
};
// @lc code=end

File: 540.有序数组中的单一元素.py
#
# @lc app=leetcode.cn id=540 lang=python3
#
# [540] 有序数组中的单一元素
#

# @lc code=start
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = left + right >> 1
            if mid % 2:
                if nums[mid] == nums[mid - 1]:
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                if nums[mid] == nums[mid + 1]:
                    left = mid + 2
                else:
                    right = mid
        return nums[left]
# @lc code=end

